src/nix-find-roots/lib/find-roots.cc

```diff
@@ -0,0 +1,233 @@
+/**
+ * @file
+ *
+ * A very simple utility to trace all the gc roots through the file-system
+ * The reason for this program is that tracing these roots is the only part of
+ * Nix that requires to run as root (because it requires reading through the
+ * user home directories to resolve the indirect roots)
+ *
+ * This program intentionnally doesnt depend on any Nix library to reduce the attack surface.
+ */
+
+#include <regex>
+#include <unistd.h>
+#include <vector>
+#include <algorithm>
+#include <fstream>
+#include <optional>
+
+#include "find-roots.hh"
+
+
+namespace nix::roots_tracer {
+namespace fs = std::filesystem;
+using std::set, std::string;
+
+static string quoteRegexChars(const string & raw)
+{
+    static auto specialRegex = std::regex(R"([.^$\\*+?()\[\]{}|])");
+    return std::regex_replace(raw, specialRegex, R"(\$&)");
+}
+static std::regex storePathRegex(const fs::path storeDir)
+{
+    return std::regex(quoteRegexChars(storeDir) + R"(/[0-9a-z]+[0-9a-zA-Z\+\-\._\?=]*)");
```

Comment:
```suggestion
    return std::regex(quoteRegexChars(storeDir) + R"(/[0-9a-z]+-(?!\.\.?(-|$))[0-9a-zA-Z\+\-\._\?=]+)");
```

See https://github.com/NixOS/nix/pull/9867/files#diff-bdf530c7e47ccaf7f924b90d1f61c3aa7866c63a3ce5ebdcbe07960284dca223R11